### ISSUE TITLE
Support for dynamic `BLOCK_COUNT`

### ISSUE THREAD

**Issue by jonas-schievink** [nickray/littlefs2#2]
Currently the user has to statically know the device size since `BLOCK_COUNT` is an associated constant. It would be useful if it would be possible to dynamically determine the device size instead.

**Comment by nickray** [comment]
I agree, `mount` should read it from the metadata (where I believe it's stored), and `format` should take it as parameter.

**Comment by nickray** [comment]
Missing upstream: https://github.com/ARMmbed/littlefs/issues/349

**Comment by TheTechRobo** [comment]
Is it possible to pass the block count to the function at runtime if *I* know it, even if littlefs doesn't? Currently it seems I have to hardcode the volume size. I am developing for ESP32, and littlefs is using one partition of the flash. I can read the partition table, which gets me the length, and using that would make me feel much better than hardcoding the size.

If I've missed something please let me know!

**Comment by robin-nitrokey** [comment]
Unfortunately this is currently not possible.  We could remove all associated constants from the `Storage` trait and pass them to `Allocation::new` instead.  The only downside is that the storage implementation could be called with unexpected offsets and block sizes, but it should probably validate those anyway.

**Comment by TheTechRobo** [comment]
That seems like a good idea to me. It would allow for reading any filesystem discovered at runtime, rather than just ones the developer thought of.

It is not a serious issue in my case: I came up with a workaround where the app reads the partition table and makes sure the hardcoded values are correct, which is robust enough for me. But it would be a nice feature.

> The only downside is that the storage implementation could be called with unexpected offsets and block sizes, but it should probably validate those anyway.

Maybe constants vital to the Storage implementation (like block size) could be put in the trait? That would remove some flexibility, though.

**Comment by robin-nitrokey** [comment]
> Maybe constants vital to the Storage implementation (like block size) could be put in the trait?

Isn’t that what we have right now?

**Comment by TheTechRobo** [comment]
I meant the ones that are hardware-dependent, like block size. But rethinking it, yeah, it should probably be all or nothing, and force the Storage implementer to do validation.

**Comment by Dominaezzz** [comment]
Would a PR for this be accepted? To turn all the `const`s in the trait into non-`const` functions?

**Comment by Dominaezzz** [comment]
Oh someone has made a PR already https://github.com/trussed-dev/littlefs2/pull/101
